src/task_manager.cc

```diff
@@ -181,71 +202,92 @@ namespace detail {
 		return task_ref;
 	}
 
-	void task_manager::invoke_callbacks(task_id tid, task_type type) {
-		for(auto& cb : task_callbacks) {
-			cb(tid, type);
+	void task_manager::invoke_callbacks(const task_id tid) {
+		for(const auto& cb : task_callbacks) {
+			cb(tid);
 		}
 	}
 
-	void task_manager::add_dependency(task* depender, task* dependee, dependency_kind kind) {
+	void task_manager::add_dependency(task* depender, task* dependee, dependency_kind kind, dependency_origin origin) {
 		assert(depender != dependee);
 		assert(depender != nullptr && dependee != nullptr);
-		depender->add_dependency({dependee, kind});
+		depender->add_dependency({dependee, kind, origin});
 		execution_front.erase(dependee);
 		max_pseudo_critical_path_length = std::max(max_pseudo_critical_path_length, depender->get_pseudo_critical_path_length());
 	}
 
+	task_id task_manager::collect_execution_front(std::unique_ptr<task> reducer) {
+		// add dependencies from a copy of the front to this task
+		const auto current_front = execution_front;
+		for(task* front_task : current_front) {
+			add_dependency(reducer.get(), front_task, dependency_kind::TRUE_DEP, dependency_origin::execution_front);
+		}
+		assert(execution_front.empty());
+		return register_task_internal(std::move(reducer)).get_id();
+	}
+
+	void task_manager::set_current_epoch(const task_id epoch) {
```

Comment:
The semantic and code duplication here with `graph_generator::per_node_data::set_current_epoch` is palpable (though not new). I'm wondering if we could generalize that.

Reply:
It probably goes beyond `set_current_epoch`, we semantically duplicate the entire dependency tracking between task and command graph generation. IMO this should be part of a larger follow-up refactoring PR, though.